### PR TITLE
updated python stuff

### PR DESCRIPTION
**Pull request by ggustavs**
Changed to update current state of Python bindings

### CODE REVIEW

**Comment by wenkokke** [comment]
@KatyaKom could you review this and let us know if you're okay to merge this or if you're currently teaching?

**Comment by wenkokke** [comment]
@ggustavs Could you check if the other instructions still work with the latest vehicle and ensure that the CI tests are using the latest version as well?

File: src/chapters/4-property-driven-training.md [hunk 1]
```diff
@@ -229,48 +229,98 @@ The definition of the optimisation problem above instantiates by taking:
         \mathcal{L}_{\mathcal{S}(\mathbf{x})} = \mathcal{I} ( || f(\mathbf{x}) - f(\hat{\mathbf{x}})|| \leq \delta)
     $$
 
-# Coding Example: generating a logical loss functions for mnist-robustness in Vehicle
+# Coding Example: generating a logical loss function in Python
 
-To generate a logical loss function usable in Python, we will use Vehicle's Python bindings that come pre-installed.
+The Python bindings now ship backend-specific helpers in `vehicle_lang.loss.tensorflow`
```
Comment (MatthewDaggitt) [inline src/chapters/4-property-driven-training.md]: Can remove the `now` I think?

File: src/chapters/4-property-driven-training.md [hunk 2]
```diff
@@ -229,48 +229,98 @@ The definition of the optimisation problem above instantiates by taking:
...
+def network(x):
+    return model(x.reshape(1, 1)).reshape(1)
+
+for step in range(50):
+    optimizer.zero_grad()
 
-which generates the loss function representing the logical loss function for the property `robust`.
+    # Task loss (e.g., MSE to a target function)
+    task_loss = torch.tensor(0.0)
+    for x_val, y_val in [(0.0, 0.0), (0.5, 1.0), (1.0, 2.0)]:
```
Comment (MatthewDaggitt) [inline src/chapters/4-property-driven-training.md]: I think it would be easier to understand if you split this out into a top-level declaration called `dataset` which you pre-cast to a `torch.tensor`? Same for the Tensorflow example below.

File: src/chapters/4-property-driven-training.md [hunk 3]
```diff
@@ -229,48 +229,98 @@ The definition of the optimisation problem above instantiates by taking:
...
+constraint_loss = spec["output_bounded"]  # name of a @property in the spec
+model = torch.nn.Sequential(
+    torch.nn.Linear(1, 8), torch.nn.ReLU(), torch.nn.Linear(8, 1)
 )
+optimizer = torch.optim.Adam(model.parameters(), lr=1e-2)
 
-robust_loss_fn = spec["robust"]
-```
+def network(x):
+    return model(x.reshape(1, 1)).reshape(1)
```
Comment (MatthewDaggitt) [inline src/chapters/4-property-driven-training.md]: Hmm this is a bit unfortunate. Why do we need this declaration? Can we not either change the declaration of the network in the .vcl file or change the declaration of model above to match?

File: src/chapters/4-property-driven-training.md [hunk 4]
```diff
@@ -229,48 +229,98 @@ The definition of the optimisation problem above instantiates by taking:
...
+The Python bindings now ship backend-specific helpers in `vehicle_lang.loss.tensorflow`
+and `vehicle_lang.loss.pytorch`. They both expose a single entry point
+`load_specification(path, *, logic=..., samplers=..., declarations=..., declaration_context=...)`
+which compiles your `.vcl` file to a dictionary of callable Python loss functions.
 
-We open a new Python file and write:
+- `logic` defaults to `DifferentiableLogic.DL2`, but you can also request `DifferentiableLogic.Vehicle`.
+- `samplers` lets you override the search strategy for Vehicle `search` expressions.
+  If omitted, a default FGSM-based sampler is used (see `DefaultTensorFlowSampler` and
+  `DefaultPyTorchSampler`).
```
Comment (MatthewDaggitt) [inline src/chapters/4-property-driven-training.md]: Hmm it's a bit weird to document all the options above but only provide explanations for two of them. To be honest, I would probably leave all the options out here, and have a section at the end that discusses the optional parameters (which can then be expanded as we see fit).

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Looks great, thank you! A few minor comments.

File: src/chapters/4-property-driven-training.md (hunk 2 above)
Comment (ggustavs) [reply src/chapters/4-property-driven-training.md]: Just made a simpler input, should retain the value of the explanation without weird syntax

File: src/chapters/4-property-driven-training.md (hunk 3 above)
Comment (ggustavs) [reply src/chapters/4-property-driven-training.md]: Yeah, entirely removed that part - my bad

File: src/chapters/4-property-driven-training.md (hunk 4 above)
Comment (ggustavs) [reply src/chapters/4-property-driven-training.md]: Just switched to documenting the important bits and referring to Read The Docs for the rest - avoids having more places to update when changing the API. The stuff I describe now should last for a longer time.

**Comment by ggustavs** [comment]
slight oopsie on reusing the same command for the commit - I had a typo and then I pressed up to redo it, which, ofc, jumped to the previous successful command :D

**Review by wenkokke — APPROVED** [review]

**Review by MatthewDaggitt — COMMENTED** [review]
Looks good to me!
